### PR TITLE
Consistently check public code permissions

### PR DESCRIPTION
**Pull request by ryanslade**
This change consistently checks whether a user can add public or private code
by first looking at site config and only falling back to user specific tags if
the user is logged in or exists.

The `CurrentUserAllowedExternalServices` method was moved from
`frontend/internal/search/repos` into the database package.

If a user is not logged in or is invalid we always fall back to the site level
setting.

In the web code we also check for the `AllowUserExternalServicePrivate` tag so
that when we start opting users in we don't hide the new UX elements.

Closes: https://github.com/sourcegraph/sourcegraph/issues/19931

### CODE REVIEW

File: internal/database/users.go [hunk 1]
```diff
@@ -1126,3 +1128,67 @@ func (u *UserStore) HasTag(ctx context.Context, userID int32, tag string) (bool,
 	}
 	return false, nil
 }
+
+// Tags returns a map with all the tags currently belonging to the user.
+func (u *UserStore) Tags(ctx context.Context, userID int32) (map[string]struct{}, error) {
```
Comment (unknwon) [inline internal/database/users.go:1133]: nit: Although using `struct{}` is often nice, is it actually making the caller have a hard time here?

The following would have a better UX for callers IMO.
```go
tags, err := u.Tags(ctx, userID)
...

if tags["MyTag"] {
	...
}
```

**Review by unknwon — APPROVED** [review]

File: internal/database/users.go (hunk 1 above)
Comment (ryanslade) [reply internal/database/users.go:1133]: Good point, I just always use `struct{}` out of habit. The only real advantage is fewer allocations and it's not like this is in a tight loop or something :) 

I'll switch to `bool`

File: cmd/frontend/graphqlbackend/external_services.go [hunk 2]
```diff
@@ -48,7 +47,10 @@ func (r *schemaResolver) AddExternalService(ctx context.Context, args *addExtern
 	// 🚨 SECURITY: Only site admins may add external services if user mode is disabled.
 	namespaceUserID := int32(0)
 	isSiteAdmin := backend.CheckCurrentUserIsSiteAdmin(ctx) == nil
-	allowUserExternalServices := searchrepos.CurrentUserAllowedExternalServices(ctx)
+	allowUserExternalServices, err := database.Users(r.db).CurrentUserAllowedExternalServices(ctx)
+	if err != nil {
+		return nil, err
+	}
```
Comment (indradhanush) [inline cmd/frontend/graphqlbackend/external_services.go:53]: Nit: A new line at the end of this code block would help separate the context of this `if` block with the next. :slightly_smiling_face:

File: cmd/frontend/internal/search/repos/repo_groups.go [hunk 3]
```diff
@@ -61,7 +61,11 @@ func ResolveRepoGroups(ctx context.Context, settings *schema.Settings) (groups m
 		groups[name] = repos
 	}
 
-	if CurrentUserAllowedExternalServices(ctx) == conf.ExternalServiceModeDisabled {
+	mode, err := database.GlobalUsers.CurrentUserAllowedExternalServices(ctx)
+	if err != nil {
+		return groups, err
+	}
+	if mode == conf.ExternalServiceModeDisabled {
```
Comment (indradhanush) [inline cmd/frontend/internal/search/repos/repo_groups.go:68]: Nit: Newline. Same as above. :smile:

File: internal/repos/syncer.go [hunk 4]
```diff
@@ -223,14 +223,15 @@ func (s *Syncer) SyncExternalService(ctx context.Context, tx *Store, externalSer
...
+	// "AllowUserExternalServicePrivate" tag, user added external services should
+	// only sync public code.
+	if isUserOwned {
+		mode, err := database.UsersWith(tx).UserAllowedExternalServices(ctx, svc.NamespaceUserID)
 		if err != nil {
-			return errors.Wrap(err, "checking user tag")
+			return errors.Wrap(err, "checking if user can add private code")
 		}
-		if !ok {
+		if mode != conf.ExternalServiceModeAll {
```
Comment (indradhanush) [inline internal/repos/syncer.go:234]: Nit: Newline. :stuck_out_tongue:

**Review by indradhanush — COMMENTED** [review]
Just nit picks. :sweat_smile:

**Comment by unknwon** [comment]
@indradhanush I like your nits 😄

**Pull request merged**